style.display setter should be fast

Constellation opened the pull request:
#### 373b5f1310c1d92cb0aef4b501a3bef651d07bf1
<pre>
style.display setter should be fast
<a href="https://bugs.webkit.org/show_bug.cgi?id=265009">https://bugs.webkit.org/show_bug.cgi?id=265009</a>
<a href="https://rdar.apple.com/118548941">rdar://118548941</a>

Reviewed by Chris Dumez.

While `style.display` setter can be called frequently, this isn&apos;t optimized well.
This patch optimizes it in three ways.

1. We should have CSSParserFastPaths for display with simple one keyword, like, `style.display = &quot;none&quot;`.
2. We found that JSCSSStyleDeclaration::put is completely disabling IC and using the slowest path.
   Since JSCSSStyleDeclaration does not have any named getters / setters, we should use JSObject::put when
   we access to named property. We add a path which uses JSObject::put when propertyName is not index.
3. CSSParserContext is frequently created. We should use bitfields to make it much smaller (from 110 =&gt; 64).

* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GeneratePut):
(InstanceOverridesIndexedDefineOwnProperty):
(InstanceOverridesNamedDefineOwnProperty):
(InstanceOverridesDefineOwnProperty):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::parseDisplay):
(WebCore::CSSParserFastPaths::maybeParseValue):

Canonical link: <a href="https://commits.webkit.org/270914@main">https://commits.webkit.org/270914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c0086020192cfc2a4f0ecb9636d6a7698903404e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26769 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5384 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28010 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28981 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24474 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27214 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2779 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24376 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4207 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22981 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3693 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3737 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23974 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29465 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24421 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24374 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29990 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3771 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27895 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5221 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4240 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3467 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4124 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/bindings/scripts/test/JS/JSTestIndexedSetterThrowingException.cpp:
```diff
@@ -219,6 +219,9 @@ bool JSTestIndexedSetterThrowingException::put(JSCell* cell, JSGlobalObject* lex
     }
 
     throwScope.assertNoException();
+    if (!parseIndex(propertyName))
```
If parseIndex() returned a valid index, we would have taken the branch at line 214 and returned early at line 218.
Therefore, it seems like this condition here should always be true.

Constellation replied:
Fixed.

cdumez commented on Source/WebCore/bindings/scripts/test/JS/JSTestIndexedSetterNoIdentifier.cpp:
```diff
@@ -219,6 +219,9 @@ bool JSTestIndexedSetterNoIdentifier::put(JSCell* cell, JSGlobalObject* lexicalG
     }
 
     throwScope.assertNoException();
+    if (!parseIndex(propertyName))
```
Same comment as below.

Constellation replied:
Fixed.

cdumez commented on Source/WebCore/bindings/scripts/test/JS/JSTestIndexedSetterWithIdentifier.cpp:
```diff
@@ -228,6 +228,9 @@ bool JSTestIndexedSetterWithIdentifier::put(JSCell* cell, JSGlobalObject* lexica
     }
 
     throwScope.assertNoException();
+    if (!parseIndex(propertyName))
```
Same comment as above.

Constellation replied:
Fixed.

cdumez commented on Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
```diff
@@ -2749,6 +2749,9 @@ bool JSTestObj::put(JSCell* cell, JSGlobalObject* lexicalGlobalObject, PropertyN
     auto throwScope = DECLARE_THROW_SCOPE(lexicalGlobalObject->vm());
 
     throwScope.assertNoException();
+    if (!parseIndex(propertyName))
```
Here the check isn't redundant though it is not clear to me why we only want to flow through if the property is an index.

Constellation replied:
This is critical role of this patch because not using `JSObject::put` means "we give up any IC optimization". We should not do that until we actually need to do. And in this case, we should use `JSObject::put`.

cdumez commented on Source/WebCore/css/parser/CSSParserContext.h:
```diff
@@ -62,42 +62,42 @@ struct CSSParserContext {
     CSSParserMode mode { HTMLStandardMode };
     std::optional<StyleRuleType> enclosingRuleType;
-    bool isHTMLDocument { false };
+    bool isHTMLDocument : 1 { false };
```
This is a temporary stack object, does the size matter that much?

Constellation replied:
Yeah, but it is showing up as a sample sometimes. Anyway, I think it is good not to make it super large.

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -901,6 +901,45 @@ static RefPtr<CSSValue> parseSimpleTransform(CSSPropertyID propertyID, StringVie
     return parseSimpleTransformList(string.characters16(), string.length());
 }
 
+static RefPtr<CSSValue> parseDisplay(StringView string)
```
I would add a comment in `consumeDisplay` to make sure we add any future single keywords here too.

Constellation replied:
Added.

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -901,6 +901,45 @@ static RefPtr<CSSValue> parseSimpleTransform(CSSPropertyID propertyID, StringVie
+    case CSSValueListItem:
+        return CSSPrimitiveValue::create(value);
+    default:
+        return nullptr;
```
I would handle CSS wide keywords here: `display: inherit / initial / revert / revert-layer`. 

```
if (isCSSWideKeyword(value))
    return CSSPrimitiveValue::create(value);
```

`parseKeywordValue` does this for free, but a special case in `parseDisplay` is probably fine too.

Constellation replied:
Sounds good. Changed.

cdumez reviewed (approved):
LGTM

Constellation committed:
style.display setter should be fast
https://bugs.webkit.org/show_bug.cgi?id=265009
rdar://118548941

Reviewed by Chris Dumez.

While `style.display` setter can be called frequently, this isn't optimized well.
This patch optimizes it in three ways.

1. We should have CSSParserFastPaths for display with simple one keyword, like, `style.display = "none"`.
2. We found that JSCSSStyleDeclaration::put is completely disabling IC and using the slowest path.
   Since JSCSSStyleDeclaration does not have any named getters / setters, we should use JSObject::put when
   we access to named property. We add a path which uses JSObject::put when propertyName is not index.
3. CSSParserContext is frequently created. We should use bitfields to make it much smaller (from 110 => 64).

* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GeneratePut):
(InstanceOverridesIndexedDefineOwnProperty):
(InstanceOverridesNamedDefineOwnProperty):
(InstanceOverridesDefineOwnProperty):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::parseDisplay):
(WebCore::CSSParserFastPaths::maybeParseValue):

Canonical link: https://commits.webkit.org/270914@main
